pypfopt/efficient_semivariance.py

```diff
@@ -0,0 +1,134 @@
+"""
+The ``efficient_semivariance`` module houses the EfficientSemivariance class, which
+generates optimal portfolios with respect to the semivariance of the portfolio.
+"""
+
+import warnings
+
+import cvxpy as cp
+import numpy as np
+import pandas as pd
+
+from . import base_optimizer, objective_functions
+
+
+class EfficientSemivariance(base_optimizer.BaseConvexOptimizer):
+
+    # Note: expected_returns and historic_returns have to be in the same frequency
+    def __init__(
+        self,
+        expected_returns,
+        historic_returns,
+        frequency=252,
+        benchmark=0,
+        weight_bounds=(0, 1),
+        solver=None,
+        verbose=False,
+    ):
+        # Inputs
+        self.expected_returns = EfficientSemivariance._validate_expected_returns(
+            expected_returns
+        )
+        self.historic_returns = historic_returns
+        self.frequency = frequency
+        self.benchmark = benchmark
+        self.T = self.historic_returns.shape[0]
+
+        # Labels
+        if isinstance(expected_returns, pd.Series):
+            tickers = list(expected_returns.index)
+        elif isinstance(historic_returns, pd.DataFrame):
+            tickers = list(historic_returns.columns)
+        else:  # use integer labels
+            tickers = list(range(len(expected_returns)))
+
+        if expected_returns is not None:
+            if historic_returns.shape[1] != len(expected_returns):
+                raise ValueError(
+                    "Historic return matrix does not match expected returns"
+                )
+
+        super().__init__(
+            len(tickers), tickers, weight_bounds, solver=solver, verbose=verbose
+        )
+
+    @staticmethod
+    def _validate_expected_returns(expected_returns):
```

Comment:
These validation functions are duplicated from the `EfficientFrontier` class. Perhaps it would make sense to define them somewhere else and reuse them in both places.